KIPs/kip-247.md

```diff
@@ -0,0 +1,186 @@
+Upon detection of GaslessTx, the block proposer may lend the user KAIA to facilitate the GaslessTx execution.
+A gasless swap from Token to KAIA shall pay back the lent amount.
+
+## Specification
+
+### Definition of GaslessTx
+
+There are two types of GaslessTx: GaslessApproveTx, and GaslessSwapTx.
+
+#### GaslessApproveTx
+
+A GaslessApproveTx must meet all the following conditions to be inserted to `txpool.queue`:
+
+- A1: `GaslessApproveTx.to` is a whitelisted ERC-20 token.
+- A2: `GaslessApproveTx.data` is `approve(spender, amount)`.
+- A3: `spender` is a whitelisted `GaslessSwapRouter`.
+- A4: `amount` is a nonzero.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A GaslessApproveTx must meet all the following conditions to be promoted to `txpool.pending`:
+
+- AP1: GaslessApproveTx is followed by a GaslessSwapTx of the same sender and the `GaslessSwapTx` can be promoted to `txpool.pending`.
+
+#### GaslessSwapTx
+
+A `GaslessSwapTx` must meet all the following conditions to be inserted to `txpool.queue`:
+
+- S1: `GaslessSwapTx.to` is a whitelisted `GaslessSwapRouter`.
+- S2: `GaslessSwapTx.data` is `swapForGas(token, amountIn, amountOut, amountRepay)`.
+- S3. `token` is a whitelisted ERC20 token.
+
+(Note that above statements can be verified solely through a static validation of a transaction.)
+
+A `GaslessSwapTx` must meet all the following conditions to be promoted to `txpool.pending`:
+
+- SP1: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.to=token`.
+- SP2: (If GaslessApproveTx exists in `txpool.queue`) `GaslessApproveTx.data.amount>=amountIn`.
+- SP3: `amountRepay` is the correct value.
+  - (If GaslessApproveTx exists in `txpool.queue`) `amountRepay = CalcRepayAmount(GaslessApproveTx, GaslessSwapTx)`.
+  - (otherwise) `amountRepay = amountRepay(GaslessSwapTx)`.
+  - `amountRepay(GaslessApproveTx, GaslessSwapTx) = R1 + R2 + R3` and `amountRepay(GaslessSwapTx) = R1 + R3` where
+    - `R1 = LendTx.Fee() = SwapTx.GasPrice() * 21000`
+    - `R2 = ApproveTx.Fee()` (if exists)
+    - `R3 = SwapTx.Fee()`
+
+While GaslessApproveTx must be followed by with GaslessSwapTx in order to be promoted, GaslessSwapTx can exist by itself.
+
+### Definition of LendTx
+
+Upon detection of a GaslessTx (either GaslessApproveTx or GaslessSwapTx), the block proposer shall inject a transaction, denoted by LendTx, which transfers KAIA to the GaslessTx sender. LendTx must meet all the following requirements:
+
+- L1: `LendTx.type` is 0x7802, a Ethereum dynamic fee transaction type.
+- L2: `LendTx.from` is the proposer itself.
+- L3: `LendTx.to` is `GaslessSwapTx.from`
+- L4: `LendTx.value` is `amountLend`
+  - `amountLend` is `R2 + R3` (from aforementioned `amountRepay` definition).
+
+### GaslessSwapRouter
+
+GaslessSwapRouter is a smart contract responsible for token swap and repayment. It must support the following interface:
```

Comment:
If we're going to have (token -> dex) mapping, please clarify that this Router only supports single-hop swap paths.

Reply:
fixed 0d1638b